packages/gateway/src/durable-objects/metrics.js

```diff
@@ -0,0 +1,150 @@
+/**
+ * @typedef {Object} GatewayMetrics
+ * @property {number} totalReqCount total number of performed requests
+ * @property {number} failedReqCount total number of requests failed
+ * @property {number} fasterReqCount number of performed requests where faster
+ * @property {Record<string, number>} responseTimeHistogram
+ * @property {number} [averageResponseTime]
+ *
+ * @typedef {Object} ResponseStats
+ * @property {string} url gateway URL
+ * @property {boolean} ok request was successful
+ * @property {number} [responseTime] number of milliseconds to get response
+ * @property {boolean} [faster]
+ */
+
+/**
+ * Durable Object for keeping Metrics state.
+ */
+export class Metrics8 {
+  constructor(state, env) {
+    this.state = state
+    /** @type {Array<string>} */
+    this.ipfsGateways = JSON.parse(env.IPFS_GATEWAYS)
+
+    // `blockConcurrencyWhile()` ensures no requests are delivered until initialization completes.
+    this.state.blockConcurrencyWhile(async () => {
+      /** @type {Map<string, GatewayMetrics>} */
+      this.gatewayMetrics = new Map()
+
+      // Get state and initialize if not existing
+      const storedMetricsPerGw = await Promise.all(
+        this.ipfsGateways.map(async (gw) => {
+          /** @type {GatewayMetrics} */
+          const value = (await this.state.storage.get(gw)) || {
+            ...defaultCounter,
+          }
+
+          return {
+            gw,
+            value,
+          }
+        })
+      )
+
+      storedMetricsPerGw.forEach(({ gw, value }) => {
+        this.gatewayMetrics.set(gw, value)
+      })
+    })
+  }
+
+  // Handle HTTP requests from clients.
+  async fetch(request) {
+    // Apply requested action.
+    let url = new URL(request.url)
+    switch (url.pathname) {
+      case '/update':
+        const data = await request.json()
+        // Get updated Metrics
+        const updatedMetrics = this._getUpdatedMetrics(data)
```

Comment:
Has side effect of updating `this.gatewayMetrics` but also returns it...personally I'd either not return anything and rename it `_updateMetrics`  or have a function that does not mutate the current metrics but creates a new set of metrics based on the last and returns them:

```js
this.gatewayMetrics = updateMetrics(this.gatewayMetrics, data)
```

Reply:
Yeah, let's not return anything. Otherwise we add more complexity by not mutating it